napari/_vispy/vendored/volume.py

```diff
@@ -362,20 +364,34 @@
     // datasets. Ugly, but it works ...
     vec3 loc = start_loc;
     int iter = 0;
+    
+    // Keep track of whether texture has been sampled
+    int texture_sampled = 0;
+    
     while (iter < nsteps) {
         for (iter=iter; iter<nsteps; iter++)
         {
-            // Get sample color
-            vec4 color = $sample(u_volumetex, loc);
-            float val = color.r;
-
-            $in_loop
-
+            // Only sample volume if loc is not clipped by clipping planes
+            float is_shown = $clip_by_planes(loc, u_shape);
+            if (is_shown >= 0)
+            {
+                // Get sample color
+                vec4 color = $sample(u_volumetex, loc);
+                float val = color.r;
+                texture_sampled = 1;
+
+                $in_loop
+            }
             // Advance location deeper into the volume
             loc += step;
         }
     }
-
+    
+    // discard fragment if texture not sampled
+    if ( texture_sampled != 1 ) {
+        discard;
+    }
+    
```

Comment:
@kevinyamauchi for reference, these are the lines we are talking about over at #3181 regarding the `discard` call.